scripts: gen_relocate_app.py: Support relocating noinit sections

pdgendt opened the pull request:
This PR adds support for relocating `noinit` sections with the `zephyr_code_relocate` cmake extension.

pdgendt committed:
scripts: gen_relocate_app.py: Support relocating noinit sections

This commit adds support for relocating noinit sections with the
zephyr_code_relocate cmake extension.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
toolchain: gcc: Allow generated unique ID as section name

Unique section names use the __FILE__ predefined MACRO,
which can contain slashes and dots. However this has
issues when trying to relocate variables or functions
inside the section.

Use the generated Z_UNIQUE_FILE_ID if defined instead.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

devanlai commented:
I just wanted to voice my support for this - I needed to place a no-init structure in a specific SRAM region because the main SRAM gets zeroed by the hardware at boot (something about parity checking) - and this does pretty much exactly what I wanted.

devanlai commented:
Hmm, actually, after doing some further testing (albeit on my own internal branch, but AFAICT the changes should be mostly independent), it seems like the unique section names generated by `__noinit` and `__noinit_named()` when compiling with gcc are formatted in a way that causes issues when the linker script attempts to match them.

For example, using `__noinit` on a test file generates a section named:
```
.noinit."WEST_TOPDIR/test/src/noinit_test.c".0
```

This is carried through into the generated `linker.cmd`:
```
 _SRAM2_NOINIT_SECTION_NAME :
        {
                . = ALIGN(4);
                KEEP(*(.noinit."WEST_TOPDIR/test/src/noinit_test.c".0))
                . = ALIGN(4);
 } > SRAM2
```

I tried adjusting the script to quote and escape the section names, but the linker still doesn't seem happy to match even then:
```
 _SRAM2_NOINIT_SECTION_NAME :
        {
                . = ALIGN(4);
                KEEP(*(".noinit.\"WEST_TOPDIR/test/src/noinit_test.c\".0"))
                . = ALIGN(4);
 } > SRAM2
```

As a quick experiment, I tweaked the gcc implementation of `__in_section()` to use `__LINE__` as the middle segment instead of `__FILE__` and that does match successfully, which seems to confirm that it's an issue with the linker script not being able to match the section names with funky characters correctly.

I'm not sure if there is a quoting scheme that would work or if the unique section names simply must be generated with less problematic characters to avoid this issue.

devanlai commented:
From further testing, it seems that the script also needs to set the `(NOLOAD)` attribute on `noinit` regions. to prevent the linker from, well, trying to reserve space to load the noinit contents, which can blow up the size of a flat bin file.

pdgendt commented:
Hey @devanlai, thanks for testing this.

I've actually had a similar issue with the unique section names, and tried to fix this in #39625, but the change would've impacted the `cmake` build system too much.

I need to dig a bit deeper for the `(NOLOAD)` attribute, if I can find some spare time.

I think today is the last feature freeze day for release v3.0 so afterwards we can get this merged in.

pdgendt committed:
cmake: generate unique ID definition for each source file

Add a helper function to generate unique source file IDs and append
these to its compile time definitions.

The helper function is called on source files added with target_sources.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt commented:
@devanlai I've added commits which should fix the unique section name issue.

devanlai reviewed (commented):
Thanks for updating this. I see a couple of rough edges:

1\. `generate_unique_source_file_ids()` needs to be called on every source file.

I see you extended all of the zephyr functions for adding source files, but that doesn't cover the case where the application uses the vanilla `target_sources()` to add source files.

I was able to work around this by manually calling `generate_unique_source_file_ids()` on my source file, but I'm not sure what the cleanest solution to this is.

2\. `gen_relocate_app.py` generates `linker_sram_noinit_relocate.ld`, but it doesn't seem to get included anywhere (for example, in `cortex_m/scripts/linker.ld`).

I'm not sure what the implications of this are, since it seems `linker_relocate.ld` has the relevant fragment anyways. Or maybe the noinit fragment is supposed to be output to `linker_sram_noinit_relocate.ld` instead of `linker_relocate.ld`?

pdgendt commented:
> 1. `generate_unique_source_file_ids()` needs to be called on every source file.

This is somewhat intentional, we shouldn't touch vanilla cmake functions like `target_sources`. This way, a user has the option, while a fallback still works like before.

> 2. `gen_relocate_app.py` generates `linker_sram_noinit_relocate.ld`, but it doesn't seem to get included anywhere (for example, in `cortex_m/scripts/linker.ld`).
> 
> I'm not sure what the implications of this are, since it seems `linker_relocate.ld` has the relevant fragment anyways. Or maybe the noinit fragment is supposed to be output to `linker_sram_noinit_relocate.ld` instead of `linker_relocate.ld`?

Me neither, someone with more linker knowledge could help out here?

devanlai commented:
> > 1. `generate_unique_source_file_ids()` needs to be called on every source file.
> 
> This is somewhat intentional, we shouldn't touch vanilla cmake functions like `target_sources`. This way, a user has the option, while a fallback still works like before.

Yeah, I agree that it would be iffy to modify vanilla cmake functions. I would feel better if it were possible to generate the unique IDs uniformly for all files, but maybe there's not a good, non-invasive way to do that.

I suppose if we're okay the unique IDs not being generated uniformly for all files, just the ones where it's important, then it should be possible to tweak `zephyr_code_relocate()` to call `generate_unique_source_file_ids()` on the source files selected for relocation. That would guarantee that any file that could be relocated will have unique section IDs that can be matched correctly.

pdgendt commented:
> I suppose if we're okay the unique IDs not being generated uniformly for all files, just the ones where it's important, then it should be possible to tweak `zephyr_code_relocate()` to call `generate_unique_source_file_ids()` on the source files selected for relocation. That would guarantee that any file that could be relocated will have unique section IDs that can be matched correctly.

I tried to do that in my very first attempt, however calling `zephyr_code_relocate` outside of the source file's cmake directory (from the application directory for example) is too late to add the compile time definition.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -82,6 +82,7 @@ function(zephyr_sources)
       message(FATAL_ERROR "zephyr_sources() was called on a directory")
     endif()
     target_sources(zephyr PRIVATE ${arg})
+    generate_unique_source_file_ids(${arg})
```
please note, you are now significantly changing this command from purely wrapping vanilla CMake to actually do more than what is described:
https://github.com/zephyrproject-rtos/zephyr/blob/f3017aa787e75c77bb7deb9f4ae45d4b4e93fbc3/cmake/modules/extensions.cmake#L47-L57

i'm really not convinced we should do this.

tejlmand replied:
Secondly, this will break as soon as someone tries to use a generator expression for adding Zephyr sources, for example like this:
https://github.com/zephyrproject-rtos/zephyr/blob/f3017aa787e75c77bb7deb9f4ae45d4b4e93fbc3/samples/tfm_integration/tfm_secure_partition/dummy_partition/CMakeLists.txt#L28-L29

tejlmand reviewed (changes requested):
First, I would really like you to open an issue with the use-case you try to solve.

You propose a fix that will impact the way each and every library is built.

Also, the fix doesn't account in any way for generator expressions or interface libraries specifying source files.
CMake allows users to create an interface lib with list of source files, and any static lib that links to such interface lib will inherit this source list and compile the sources as part of itself.

So I see several risks in the proposed solution, but if you could describe exactly why / when you need to relocate init sections, and whether it's for whole system or a single pre-know objects, then it could be easier to consider possible solutions.

pdgendt commented:
Thanks @tejlmand for your response, I agree that the current proposal is too intrusive.

> First, I would really like you to open an issue with the use-case you try to solve.

I've created [an issue](https://github.com/zephyrproject-rtos/zephyr/issues/49155) describing a use-case, hopefully it makes sense 😃
